src/backend/distributed/planner/local_plan_cache.c

```diff
@@ -99,6 +98,86 @@ CacheLocalPlanForShardQuery(Task *task, DistributedPlan *originalDistributedPlan
 }
 
 
+/*
+ * GetLocalShardQuery is a helper function which generates the local shard
+ * query based on the jobQuery.
+ *
+ * It is not guaranteed to have consistent attribute numbers on the shards
+ * and on the shell (e.g., distributed/reference tables) due to DROP COLUMN
+ * commands. To avoid any edge cases due to such discrepancies, we first
+ * deparse the jobQuery with the tables replaced to shards, and parse the
+ * query string back. This is normally a very expensive operation, however
+ * we only do it once per cached local plan, which is acceptable.
+ */
+static Query *
+GetLocalShardQuery(Query *jobQuery, Task *task, ParamListInfo orig_paramListInfo)
+{
+	StringInfo queryString = makeStringInfo();
+
+	/*
+	 * We imitate what TaskQueryString() does, but we cannot rely on that function
+	 * as the parameters might have been already resolved on the QueryTree in the
+	 * task. Instead, we operate on the jobQuery where are sure that the
+	 * coordination evaluation has not happened.
+	 *
+	 * Local shard queries are only applicable for local cached query execution.
+	 * In the local cached query execution mode, we can use a query structure
+	 * (or query string) with unevaluated expressions as we allow function calls
+	 * to be evaluated when the query on the shard is executed (e.g., do no have
+	 * coordinator evaluation, instead let Postgres executor evaluate values).
+	 *
+	 * Additionally, we can allow them to be evaluated again because they are stable,
+	 * and we do not cache plans / use unevaluated query strings for queries containing
+	 * volatile functions.
+	 */
+	if (jobQuery->commandType == CMD_INSERT)
+	{
+		/*
+		 * We cannot use deparse_shard_query for INSERT .. SELECT, and we
+		 * currently do not support INSERT .. SELECT here.
```

Comment:
I don't think this statement is correct. We use deparse_shard_query in RouterModifyTaskForShardInterval for INSERT..SELECT.